File: CS/DevExpressWalkthrough/ViewModels/Supplier/SupplierViewModel.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpressWalkthrough.Common.Utils;
using DevExpressWalkthrough.NorthwindEntitiesDataModel;
using DevExpressWalkthrough.Common.DataModel;
using DevExpressWalkthrough;
using DevExpressWalkthrough.Common.ViewModel;

namespace DevExpressWalkthrough.ViewModels {
    /// <summary>
    /// Represents the single Supplier object view model.
    /// </summary>
    public partial class SupplierViewModel : SingleObjectViewModel<Supplier, int, INorthwindEntitiesUnitOfWork> {

        /// <summary>
        /// Creates a new instance of SupplierViewModel as a POCO view model.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        public static SupplierViewModel Create(IUnitOfWorkFactory<INorthwindEntitiesUnitOfWork> unitOfWorkFactory = null) {
            return ViewModelSource.Create(() => new SupplierViewModel(unitOfWorkFactory));
        }

        /// <summary>
        /// Initializes a new instance of the SupplierViewModel class.
        /// This constructor is declared protected to avoid undesired instantiation of the SupplierViewModel type without the POCO proxy factory.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        protected SupplierViewModel(IUnitOfWorkFactory<INorthwindEntitiesUnitOfWork> unitOfWorkFactory = null)
            : base(unitOfWorkFactory ?? UnitOfWorkSource.GetUnitOfWorkFactory(), x => x.Suppliers, x => x.CompanyName) {
        }

        /// <summary>
        /// The view model for the SupplierProducts detail collection.
        /// </summary>
        public CollectionViewModel<Product, int, INorthwindEntitiesUnitOfWork> SupplierProductsDetails
        {
            get { return GetDetailsCollectionViewModel((SupplierViewModel x) => x.SupplierProductsDetails, x => x.Products, x => x.SupplierID, (x, key) => { x.SupplierID = key; }); }
        }
    }
}

File: VB/DevExpressWalkthrough.VB/ViewModels/Supplier/SupplierViewModel.vb
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpressWalkthrough.VB.Common.Utils
Imports DevExpressWalkthrough.VB.NorthwindEntitiesDataModel
Imports DevExpressWalkthrough.VB.Common.DataModel
Imports DevExpressWalkthrough.VB
Imports DevExpressWalkthrough.VB.Common.ViewModel
Namespace Global.DevExpressWalkthrough.VB.ViewModels
    ''' <summary>
    ''' Represents the single Supplier object view model.
    ''' </summary>
    Public Partial Class SupplierViewModel
        Inherits SingleObjectViewModel(Of Supplier, Integer, INorthwindEntitiesUnitOfWork)
        ''' <summary>
        ''' Creates a new instance of SupplierViewModel as a POCO view model.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Public Shared Function Create(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of INorthwindEntitiesUnitOfWork) = Nothing) As SupplierViewModel
            Return ViewModelSource.Create(Function() New SupplierViewModel(unitOfWorkFactory))
        End Function
        ''' <summary>
        ''' Initializes a new instance of the SupplierViewModel class.
        ''' This constructor is declared protected to avoid undesired instantiation of the SupplierViewModel type without the POCO proxy factory.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Protected Sub New(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of INorthwindEntitiesUnitOfWork) = Nothing)
            MyBase.New(If(unitOfWorkFactory, UnitOfWorkSource.GetUnitOfWorkFactory()), Function(ByVal x) x.Suppliers, Function(ByVal x) x.CompanyName)
        End Sub
        ''' <summary>
        ''' The view model for the SupplierProducts detail collection.
        ''' </summary>
        Public ReadOnly Property SupplierProductsDetails As CollectionViewModel(Of Product, Integer, INorthwindEntitiesUnitOfWork)
            Get
                Return GetDetailsCollectionViewModel(Function(ByVal x As SupplierViewModel) x.SupplierProductsDetails, Function(ByVal x) x.Products, Function(ByVal x) x.SupplierID, Sub(ByVal x, ByVal key) x.SupplierID = key)
            End Get
        End Property
    End Class
End Namespace
